Plutarch/Lift.hs

```diff
@@ -1,147 +1,69 @@
--}
-class PLift (p :: k -> Type) where
-  -- | The associated Haskell type for `p`
-  type PHaskellType p :: Type
-
-  -- {-
-  -- Create a Plutarch-level constant, from a Haskell value.
-  -- Example:
-  -- > pconstant @PInteger 42
-  -- -}
-  pconstant' :: PHaskellType p -> Term s p
-
-  -- {-
-  -- Convert a Plutarch term to the associated Haskell value. Fail otherwise.
-  -- This will fully evaluate the arbitrary closed expression, and convert the
-  -- resulting value.
-  -- -}
-  plift' :: ClosedTerm p -> Either LiftError (PHaskellType p)
-
--- | Like `pconstant'` but TypeApplication-friendly
-pconstant :: forall p s. PLift p => PHaskellType p -> Term s p
-pconstant = pconstant'
 
--- | Like `plift'` but fails on error.
-plift :: (PLift p, HasCallStack) => ClosedTerm p -> PHaskellType p
-plift prog = either (error . show) id $ plift' prog
+-- FIXME: `h -> p`
+class (PLifted p ~ h, PLC.DefaultUni `PLC.Includes` PLiftedRepr p) => PUnsafeLiftDecl (h :: Type) (p :: k -> Type) | p -> h where
+  type PLiftedRepr p :: Type
+  type PLifted p :: Type
+  pliftToRepr :: h -> PLiftedRepr p
+  pliftFromRepr :: PLiftedRepr p -> Maybe h
 
---------------------------------------------------------------------------------
+class PUnsafeLiftDecl (PLifted p) p => PLift (p :: k -> Type)
+instance PUnsafeLiftDecl (PLifted p) p => PLift (p :: k -> Type)
 
-{- | DerivingVia representation to auto-derive `PLift` for Plutarch types
- representing builtin Plutus types in the `DefaultUni`.
+newtype DerivePLiftViaCoercible (h :: Type) (p :: k -> Type) (r :: Type) (s :: k) = DerivePLiftViaCoercible (p s)
 
- The `h` parameter is the Haskell type associated with `p`. Example use:
+instance (Coercible h r, PLC.DefaultUni `PLC.Includes` r) => PUnsafeLiftDecl h (DerivePLiftViaCoercible h p r) where
+  type PLiftedRepr (DerivePLiftViaCoercible h p r) = r
+  type PLifted (DerivePLiftViaCoercible h p r) = h
+  pliftToRepr = coerce
+  pliftFromRepr = Just . coerce
 
- > deriving PLift via (PBuiltinType PInteger Integer)
-
- See instance below.
--}
-newtype PBuiltinType (p :: k -> Type) (h :: Type) s = PBuiltinType (p s)
-
-instance {-# OVERLAPPABLE #-} PLift (PBuiltinType p h) where
-  type PHaskellType (PBuiltinType p h) = h
+pconstant :: forall p h s. PUnsafeLiftDecl h p => h -> Term s p
+pconstant x = punsafeConstantInternal $ PLC.someValue @(PLiftedRepr p) @PLC.DefaultUni $ pliftToRepr @_ @h @p x
 
+-- | Error during script evaluation.
+data LiftError = LiftError deriving stock (Eq, Show)
+
+plift' :: forall p h. PUnsafeLiftDecl h p => ClosedTerm p -> Either LiftError h
+plift' prog = case evaluateScript (compile prog) of
+  Right (_, _, Scripts.unScript -> UPLC.Program _ _ term) ->
+    case readKnownConstant @_ @(PLiftedRepr p) @(MachineError PLC.DefaultFun) Nothing term of
+      Right r -> case pliftFromRepr @_ @h @p r of
+        Just h -> Right h
+        Nothing -> Left LiftError
+      Left _ -> Left LiftError
+  Left _ -> Left LiftError
```

Comment:
Why are we dropping error messages here (and elsewhere)?

Reply:
It was simply because I wanted to get staging to compile again ASAP.

Reply:
These will definitely be added back before merging into main, hence the FIXME.